File: LeetCode Solutions/May LeetCoding Challenges/Binary Tree Cameras.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    minNumberOfCameras = 0
    def minCameraCover(self, root: TreeNode) -> int:
        def helper(node: TreeNode) -> int:
            if(not node):
                return 0
            value = helper(node.left) + helper(node.right)
            if(value == 0):
                return 3
            if(value < 3):
                return 0
            self.minNumberOfCameras += 1
            return 1
        if(helper(root) > 2):
            return self.minNumberOfCameras + 1
        else:
            return self.minNumberOfCameras

File: LeetCode Solutions/May 2021 LeetCoding Challenges/Binary Tree Cameras/Binary Tree Cameras.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int minNumberOfCameras = 0;
    public int minCameraCover(TreeNode root) {
        return helper(root) > 2 ? minNumberOfCameras + 1 : minNumberOfCameras;
    }
    public int helper(TreeNode node) {
        if (node == null){
            return 0;
        }
        int value = helper(node.left) + helper(node.right);
        if (value == 0){
            return 3;
        }
        if (value < 3){
            return 0;
        }
        minNumberOfCameras++;
        return 1;
    }
}